### PR TITLE
Feature/sample from distribution tf

### PR DESCRIPTION
**Pull request by mstechly**
Since the previous version of `sample_from_distribution` was not working with tf backend I added my version. 
Since there are two separate versions of `sample_from_distribution` for numerical and tf, I deleted `helpers.py` and this function into appropriate files.

### CODE REVIEW

**Comment by co9olguy** [comment]
@mariaschuld what do you think of this PR?

@mstechly I notice that the change causes test failures (probably because you moved the location of the function). Can you update the PR to make the tests pass? 

Also, I believe this function is employed in the tutorials in the docs. We would also need to update those and ensure they run as expected

**Comment by co9olguy** [comment]
@mstechly Great thanks! :smile: 
Going to assign @mariaschuld to review this one. If she decides to merge, we'll also have to update the docs, as this function appears there as well:
https://qmlt.readthedocs.io/en/latest/tutorials/tensorflow.html?highlight=sample_from_distr
https://qmlt.readthedocs.io/en/latest/tutorials/tensorflow.html?highlight=sample_from_distr

**Comment by mariaschuld** [comment]
Will merge and update the documentation as soon as I am back from vacation!

File: examples/tf/unsupervised_advanced_tf.py [hunk 1]
```diff
@@ -105,7 +105,9 @@ def myregularizer(regularized_params):
 final_distribution = outcomes['outputs']
 
 # Use a helper function to sample fock states from this state.
-# They should show a similar distribution to the training data
+# They should show a similar distribution to the training data.
+# If the output of the function is a tensor, 
+# you can use sample_from_distribution from qmlt.tf.helpers
```
Comment (mariaschuld) [inline examples/tf/unsupervised_advanced_tf.py]: Would be great if you could shorten the description. Lines 109 & 110 seem redundant to me?

File: examples/tf/unsupervised_tf.py [hunk 2]
```diff
@@ -91,7 +91,9 @@ def myregularizer(regularized_params):
 final_distribution = outcomes['outputs']
 
 # Use a helper function to sample fock states from this state.
-# They should show a similar distribution to the training data
+# They should show a similar distribution to the training data.
+# If the output of the function is a tensor, 
+# you can use sample_from_distribution from qmlt.tf.helpers
```
Comment (mariaschuld) [inline examples/tf/unsupervised_tf.py]: See before.

File: qmlt/tf/helpers.py [hunk 3]
```diff
@@ -121,3 +121,27 @@ def make_param(name=None, stdev=None, mean=0., interval=None, constant=None, sha
...
+
+    probs_flat = tf.reshape(distribution, [-1])
+    rescaled_probs = tf.expand_dims(tf.log(probs_flat), 0)
+    indices_flat = tf.range(probs_flat.shape[0])
+    indices = tf.reshape(indices_flat, [cutoff] * num_modes)
+
+    sample_index = tf.squeeze(tf.multinomial(rescaled_probs, 1),[0])
+    fock_state = tf.reshape(tf.where(tf.equal(indices, tf.cast(sample_index, dtype=tf.int32))), [-1])
+
+    return fock_state
```
Comment (mariaschuld) [inline qmlt/tf/helpers.py]: Minor detail, but python best practice wants a new line after the return statement...

File: examples/tf/unsupervised_tf.py [hunk 4]
```diff
@@ -29,7 +29,7 @@
 import tensorflow as tf
 from qmlt.tf import CircuitLearner
 from qmlt.tf.helpers import make_param
-from qmlt.helpers import sample_from_distribution
+from qmlt.numerical.helpers import sample_from_distribution
```
Comment (mariaschuld) [inline examples/tf/unsupervised_tf.py]: Why did you make a tf version of the sample_from_distribution function if you import the numerical one here? Isn't that very confusing to use a "numerical helper" in a tf example? (see Conversation tab)

**Comment by mstechly** [comment]
@mariaschuld I shortened descriptions as you asked.
Concerning the docs - I suppose there is no need for `helpers.rst` anymore. However, I don't know why `numerical.helpers.rst` and `tf.helpers.rst` are different - could you explain which version is better and why?

**Comment by mariaschuld** [comment]
Hey @mstechly! Overall looking good, but I am not entirely sure what use case you have in mind for the tf sample_from_distribution function? In particular, because I see that in the tf unsupervised tutorial you still import the numerical one, since the sampling is part of the postprocessing (when we are not dealing with tf.Tensor objects any more). 

Isn't it more confusing to rely on a helper from the numerical library rather than to import it from a higher level as before? Did you need the tf version for your purpose somewhere?

If you have a use case, we can think of putting your tf version into the tf helper and leave everything else as is. Would that make sense to you?

**Comment by mstechly** [comment]
@mariaschuld Honestly, I'm also not sure if the solution I proposed is the best one. I was thinking where to include this code and decided that this way it's better, since I don't need to import `tf` in `helpers.py`. Another solution I was thinking about was adding a flag `use_tf` to the existing `sample_from_distribution` and put `import tf` and all the logic inside the `if` when the flag is true.

You can check out my use-case here: https://github.com/BOHRTECHNOLOGY/yellow_submarine/blob/master/maxcut_solver_tf.py , line 105.

I wanted to calculate loss function based on a single output of the circuit (bitstring). But since outputs are probabilistic, I wanted to use several outputs to calculate a mean loss for a given set of parameters (similar to batch learning). I wasn't able to achieve that with:
- adding measurements at the end of the circuit and sampling several times - because then I get always the same value
- adding measurements at the end of the circuit and running the circuit several times - I got some tensorflow errors I couldn't solve.
So I decided to do the following - use `all_fock_probs` and "simulate" the measurements by sampling from the distribution.
This allows me to run circuit once (avoid tensorflow errors and actually run the whole circuit only once) and get a number of measurements.

**Comment by co9olguy** [comment]
Hi @mstechly,

I'm trying to understand what the use-case is here as well. If your goal is to calculate a mean loss value, can you not do this directly from the probabilities (i.e., take `all_fock_probs` and weight the values by these probabilities) instead of estimating the mean empirically from a few samples?

**Comment by mstechly** [comment]
@co9olguy I can, but I think it's less correct from the methodological point of view.

I'm solving MaxCut problem. If I wanted to use probabilities in conjunction with my probabilities, I had to evaluate every possible solution to calculate loss.
But it doesn't make sense - if I can calculate every possible solution, then I don't need my quantum circuit to solve it. Sure - I can do that for small circuits, but this solution won't scale for bigger graphs.
But what I can do is evaluating every sample I draw from the distribution.

**Comment by co9olguy** [comment]
Still not completely clear :thinking: 
Since you are training a model in a simulator backend, which has access to probabilities, why wouldn't you use this to direct your training? On real hardware, you would of course have to estimate these probabilties empirically. But working in simulator, I don't understand why you would throw away the available information
